iceoryx_utils/include/iceoryx_utils/concurrent/resizeable_lockfree_queue.hpp

```diff
@@ -0,0 +1,198 @@
+    using Base::tryPush;
+
+    /// @brief tries to remove value in FIFO order
+    /// @return value if removal was successful, empty optional otherwise
+    /// @note threadsafe, lockfree
+    /// iox::cxx::optional<ElementType> pop() noexcept;
+    using Base::pop;
+
+    /// @brief check whether the queue is empty
+    /// @return true iff the queue is empty
+    /// @note that if the queue is used concurrently it might
+    /// not be empty anymore after the call
+    ///  (but it was at some point during the call)
+    /// @note threadsafe, lockfree
+    /// bool empty() const noexcept;
+    using Base::empty;
+
+    /// @brief get the number of stored elements in the queue
+    /// @return number of stored elements in the queue
+    /// @note that this will not be perfectly in sync with the actual number of contained elements
+    /// during concurrent operation but will always be at most capacity
+    /// @note threadsafe, lockfree
+    /// uint64_t size() const noexcept;
+    using Base::size;
+
+    /// @brief returns the current capacity of the queue
+    /// @return the current capacity
+    /// @note threadsafe, lockfree
+    uint64_t capacity() const noexcept;
+
+    /// @brief inserts value in FIFO order, always succeeds by removing the oldest value
+    /// when the queue is detected to be full (overflow)
+    /// @param[in] value to be inserted is copied into the queue
+    /// @return removed value if an overflow occured, empty optional otherwise
+    /// @note threadsafe, lockfree
+    iox::cxx::optional<ElementType> push(const ElementType& value) noexcept;
+
+    /// @brief inserts value in FIFO order, always succeeds by removing the oldest value
+    /// when the queue is detected to be full (overflow)
+    /// @param[in] value to be inserted is moved into the queue if possible
+    /// @return removed value if an overflow occured, empty optional otherwise
+    /// @note threadsafe, lockfree
+    iox::cxx::optional<ElementType> push(ElementType&& value) noexcept;
+
+    // multiple overloads to set the capacity
+    // 1) The most general one allows providing a removeHandler to specify remove behavior.
+    // 2) The overload where a container is provided can be used if the removed elements are to be stored.
+    //    The container must satisfy certain requirements, such as providing push_back.
+    // 3) The final overload discards removed elements.
+
+    /// @brief      Set the capacity to some value.
+    /// @param[in]  newCapacity capacity to be set
+    /// @param[in]  removeHandler is a function taking an element which specifies
+    ///             what to do with removed elements should the need for removal arise.
+    /// @return     true if the capacity was successfully set, false otherwise
+    template <typename Function,
+              typename = typename std::enable_if<cxx::is_invocable<Function, ElementType>::value>::type>
+    bool setCapacity(uint64_t newCapacity, Function&& removeHandler) noexcept;
```

Comment:
I commented this for above. It is no good style for value parameters. I can do it, but would prefer a better argument other that some AUTOSAR rules demands it (or none at all). Where is the advantage? Why is it better?

Reply:
> Where is the advantage? Why is it better?

1. Performance: https://www.reddit.com/r/cpp/comments/5gktfc/please_declare_your_variables_as_const/dat2rcb/

2. Clear Intent: https://dev.to/fenbf/please-declare-your-variables-as-const

3. Clean Code: https://www.bfilipek.com/2016/11/iife-for-complex-initialization.html

4. Fewer Bugs: https://akrzemi1.wordpress.com/2016/12/16/more-const-fewer-bugs/

5. const correctness is well accepted in the C++ community, see:
     - [CPP Core Guidelines](https://www.modernescpp.com/index.php/c-core-guidelines-rules-for-constants-and-immutability)
     - Scott Meyer, Effective C++, Chapter 3

6. It is our coding guideline.

Reply:
I am aware of all of them but the details of optimization, which are quite intricate.

1. Performance is a good reason, the scenarios are few however from what I see and I would suspect that it is not quite as simple for parameters, see below.
2. sure for local variables, but not for function parameters of value type
3. and 4. refer to local variables (other than parameters), which is not the case here

Note that for internal implementations I fully agree. Just not for interfaces because it may lead to more variables (to copy the value) internally as I stated which in all likelihood will make it not faster (these have to be created and live in registers/memory at least which means there will be additional instructions if the orginal one cannot be modified).

Clear intent for interfaces I refute, the user should not be concerned about this and it does not affect the caller in any way as a copy is made with or without const.

To summarize: my rule would be: 
i) make references and pointers const where everpossible
ii) make local variables const if the implementation allows it 
iii) never make value parameters const
iv) make global read only variables const if possible (i.e. global constants)

Reply:
Additional reading material:
https://google.github.io/styleguide/cppguide.html#Use_of_const
and a somewhat lengthy discussion which mostly tends to not using const in those cases as well
https://stackoverflow.com/questions/117293/use-of-const-for-function-parameters

Note clang tidy even has an option ``readability-avoid-const-params-in-decls``, which indicates it is considered bad style by many and we should not use this style (the optimization argument will usually not apply, we have no compile time constants involved if it is a parameter).

Reply:
AUTOSAR does not seem to require it in https://www.autosar.org/fileadmin/user_upload/standards/adaptive/18-10/AUTOSAR_RS_CPP14Guidelines.pdf (section 6.7.1  Specifiers). If we want to do this, it is essentially our own rule.